### PR TITLE
allow all access levels from icatDefines.h

### PR DESCRIPTION
**Pull request by erwinvanwieringen**
This will allow all access permissions from icatDefines.h to be used when setting permissions. Since 'write' and 'read' are still valid as permission, this will be fully backward compatible.
I had some doubt about the spaces in access permission names, but it feels a bit silly to translate them all to new names without spaces. So I left them as they are.

This will work with all clients that depend on the irods api to check if a specified access permission is valid. Tested with ichmod, msiSetAcl, and the python-irodsclient.

Example usage: `ichmod "modify metadata" 'user1' /tempZone/data`

Resolves irods/irods#844

### CODE REVIEW

**Review by alanking — CHANGES_REQUESTED** [review]
First of all, thanks for the contribution!

Before merging, we will need to think through the implications of this change, but this definitely looks like the right direction.

It would be good if we could get some automated tests for the newly introduced options in order to instill confidence in the changes. One of us on the iRODS team can do this, or you can try to add some tests if you'd like. :) I think `test_ichmod.py` would be an appropriate place to add these. Here is an example for how this could be started:
```python
    def test_ichmod_modify_metadata(self):
        filename = 'test_ichmod_modify_metadata'
        local_file_path = os.path.join(self.admin.local_session_dir, filename)
        logical_path = os.path.join('/' + self.admin.zone_name, 'home', 'public', filename)

        # make a local file
        lib.make_file(local_file_path, 1)

        # put it as a data object in a collection where another user will have permissions
        self.admin.assert_icommand(['iput', local_file_path, logical_path])

        # ensure that the user cannot do anything with the data object
        self.user.assert_icommand_fail(['iget', logical_path, '-'], 'STDERR', 'CAT_NO_ACCESS_PERMISSION')

        # change permissions to "modify metadata"
        self.admin.assert_icommand(['ichmod', 'modify metadata', self.user.username, logical_path])

        # ensure that the user can modify metadata and everything below that (read, etc.)
        self.user.assert_icommand(['iget', logical_path, '-'], 'STDOUT', '0')
        # TODO: add some metadata
        # TODO: modify metadata
        # TODO: etc.

        # ensure that the user cannot do anything that requires ACL higher than "modify metadata"
        self.user.assert_icommand_fail(['iput', '-f', local_file_path, logical_path], 'STDERR', 'CAT_NO_ACCESS_PERMISSION')
        # TODO: try to irm
        # TODO: etc.
```

Also, the spaces in the permission names is actually something we are considering changing, and it would be best to do so in a .0 release (in other words, 4.3.0). The team will need to think about the implications of such a change, but we have agreed internally, at least initially, that the changes would be good to do in the same release. That said, the removal of the spaces doesn't need to happen in this commit.

File: plugins/database/src/db_plugin.cpp
Comment (erwinvanwieringen) [reply plugins/database/src/db_plugin.cpp]: I fixed this in 4cf2359. Changed `myAccesLev` to `const char*` as you suggested, and now assign a ref to a string literal form icatDefines.h

**Comment by erwinvanwieringen** [comment]
I added some tests, as you suggested. I now test a couple of operations (imeta get/set/rm, iput, iget, irm, imkdir, irmdir) with various access levels. There were some suprising results (why do you need `modify object` to create an object? You would think `create object` would suffice), but that may be caused by my interpretation of how the permission model should work. I now coded the tests so they all pass.

I added a large dictionary called `test_commands` here to configure the tests. This should be easy to extend, if you want to add extra permission related tests.

Permission levels that are defined in `icatDefines.h`, but are not in use anywhere in the code (like `ACCESS_READ_ANNOTATION`) are now not included in the tests.

File: scripts/irods/test/test_ichmod.py
Comment (alanking) [inline scripts/irods/test/test_ichmod.py:90]: Agreed that this seems wrong. This is likely an artifact of the fact that the interface for changing permissions has been the way that it is, and the reason this change exists.

**Review by alanking — COMMENTED** [review]
This is great!

We really like the framework presented in your test case. We reviewed this and found that it has illuminated some dark corners surrounding the permissions model, which is in large part why we have never bothered to look at all this stuff.

In particular, your paring down of the permissions to 10 from the original 18 seems good. The unused permissions will likely continue to be unused. We have some more conversations to have before this is settled.

And as far as writing the tests to make them pass, that is fine for now. We are of the opinion that the permission model may need to conform to what makes sense in the future, but this is good for now.

How do you feel about us pushing changes on top of your branch (currently enabled via Github)? You will receive credit for your work, but there are some things we may add as part of our process before merging.

I left a couple of other comments for your consideration. Thanks again for your work!

**Comment by erwinvanwieringen** [comment]
Break statement has been added as suggested. Also added tests for the original **read** and **write** permission levels. 
If you want to push changes on top of this branch, that is fine with me!

**Comment by alanking** [comment]
We'd like to get this change in as soon as possible, but we are working to get 4.3.0 solidified in the run up to its release. So, for now, this change looks good, but we'd like to hold off on merging it right now as we work on other tasks and evaluate the discoveries made in the permissions model. As we approach the release, we will come back to this and consider it for inclusion in 4.3.0. Thanks!

**Comment by trel** [comment]
we're looping back around to get this merged this week.

**Comment by trel** [comment]
ready for review - will # when blessed.

File: scripts/irods/test/test_ichmod.py
Comment (korydraughn) [inline scripts/irods/test/test_ichmod.py:75]: acces -> access

Comment (korydraughn) [inline scripts/irods/test/test_ichmod.py:75]: acces -> access

Comment (trel) [reply scripts/irods/test/test_ichmod.py]: done

File: scripts/irods/test/test_ichmod.py
Comment (trel) [inline scripts/irods/test/test_ichmod.py:88]: should this be READ_OBJECT?

Comment (korydraughn) [reply scripts/irods/test/test_ichmod.py:88]: I think you're correct. And `STDERR` should be removed because `ichmod` doesn't produce output on success.

Comment (trel) [reply scripts/irods/test/test_ichmod.py:88]: well, this is `iget`...

Comment (korydraughn) [reply scripts/irods/test/test_ichmod.py:88]: Whoops. You're right. Ignore me lol

**Comment by alanking** [comment]
Found an issue related to tickets while testing these changes, still investigating...

**Comment by trel** [comment]
i broke the tickets with a bad clang-format - i think we're back... hoping to merge soon.

File: scripts/irods/test/test_ichmod.py
Comment (alanking) [reply scripts/irods/test/test_ichmod.py:88]: I think this should be `Access.READ_OBJECT`, yes. This test currently passes - I *think* - because it checks for anything or nothing being on stderr, which will always be the case. So, the suggested edit by @korydraughn should be the right thing to do regardless.

Comment (trel) [reply scripts/irods/test/test_ichmod.py]: done - now incorporated.

**Comment by trel** [comment]
#'d

**Pull request merged**
